Integrate WavPack library to export/import wavpack files

Subhra264 opened the pull request:
Resolves: #1470 

*(short description of the changes and the motivation to make the changes)*
This Pull Request is created to integrate wavpack with audacity so that users can import/export wavpack files.

### Export:
1. By default, hybrid mode is not selected and the wavpack library creates a pure lossless "wv" file. If hybrid mode is enabled and the "created correction file" option is not selected then it will create a single lossy "wv" file.
![Export wavpack image 1](https://user-images.githubusercontent.com/59690052/155146822-f9a2c96e-e901-41a9-8593-8ed005e9a80f.png)

2.  Enabled "create correction file", so it will create a lossy "wv" file along with its "wvc" correction file.
![Export wavpack image 2](https://user-images.githubusercontent.com/59690052/155146997-0ce94bc2-ccf7-4f6b-94ee-1c24edb8d4bd.png)

3. Added the value for "Artist Name" tag.
![Export wavpack image 3](https://user-images.githubusercontent.com/59690052/155147166-1b2ab3cf-3652-460c-aabd-cb23cbd5671b.png)

4. 
![Export wavpack image 4](https://user-images.githubusercontent.com/59690052/155147215-8cc72364-879f-4f68-9435-7f860c335953.png)

5. 
![Export wavpack image 5 (with wvc)](https://user-images.githubusercontent.com/59690052/155147264-b56965b1-e0e7-4d10-a822-45ab1b784e4b.png)

### Import
1. 
![Import wavpack image 1(with wvc)](https://user-images.githubusercontent.com/59690052/155147359-3290ca16-cd46-47a0-b987-d630ce8c6933.png)

2. 
![Import wavpack image 2(with wvc)](https://user-images.githubusercontent.com/59690052/155147429-4a3504db-95ac-408c-b63a-7836dbbb8d32.png)

3.
![Import wavpack image 3(with wvc)](https://user-images.githubusercontent.com/59690052/155147475-8e5f78cd-1364-40da-8708-a939c17d1f4c.png)

4. Exported tags are present in the imported audio
![Import wavpack image 4(with wvc)](https://user-images.githubusercontent.com/59690052/155148234-3b4f6ed0-393c-4fb3-9eff-8c76d97ad4c1.png)

Also, the exported "wv" file successfully opens with the "VLC Media Player".

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

LWinterberg commented:
Hello @Subhra264, are you planning to continue work on this?

Subhra264 commented:
Hello, sorry, I know that I am late( I had no idea about audio channels, samples, sample rate etc. so had to study it first), but I am still working on it..

The `ImportWavPack.cpp` file is almost ready, except the `ImportWavPackFileHandle::Import` function which is partly complete. For example I pushed all the `channels` to `outTracks`. But it's taking time because I am not sure what to do with the `samplesRead` decoded with `WavpackUnpackSamples` as  in  the `ImportFLAC.cpp` file, I saw that channels were directly pushed to `outTracks`,  but in `ImportOgg.cpp` file, the read samples were appended to channels.
Also in `Import.cpp` and `ProjectFileManager.cpp`(I forgot the exact name but it was something similar), it looked like only those `tracks` were used.
Basically, I am confused that if only channels created with `WaveTrackFactory` and `sampleRate` are enough or I need to somehow append the decoded samples to the channels.

Hopefully I will be able complete in 1 or 2 days.
Thanks...

crsib commented:
Hi @Subhra264 

Sorry for the delay! Can you please rebase onto the latest master, squash you changes into a single commit and force push?

Paul-Licameli commented:
> Hi @Subhra264
> 
> Sorry for the delay! Can you please rebase onto the latest master, squash you changes into a single commit and force push?

I would like to see a rebase too, but not necessarily a squash to one commit, if the intermediate steps are really useful for review of the new feature in sensible incremental steps, and not just corrections of mistakes.

Subhra264 commented:
> Sorry for the delay!

No Issues! :)
>  but not necessarily a squash to one commit, if the intermediate steps are really useful for review of the new feature in sensible incremental steps, and not just corrections of mistakes.

yeah, actually, I added new commits after the previous commit is bug-free ( I mean, after resolving the bugs I encountered related to that particular feature/implementation, so may not be "bug-free" ). So, each commit mostly represents a new feature/file. For example, the first commit is about integrating wavpack and creating a file containing all the necessary class details, the second commit implements all the `ImportPlugin` and `ImportFileHandle` class methods, and the third commit is to read tags, etc.

Subhra264 commented:
> I have crafted a commit, that should resolve your CI issues https://github.com/crsib/audacity/commit/2a353fd357f6cebb1a9cda61be33a9afe3edfb36

All checks passed! Thank you!

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,431 @@
+
+   Subhradeep Chakraborty
+
+   This program is distributed under the GNU General Public License, version 2.
```
Please update to GPL2-or-later.

Or even better, please add `/*  SPDX-License-Identifier: GPL-2.0-or-later */` to the top of both Export and Import

Subhra264 replied:
done

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+      auto mixer = CreateMixer(tracks, selectionOnly,
+         t0, t1,
+         numChannels, SAMPLES_PER_RUN, true,
+         rate, int16Sample, mixerSpec);
```
(I believe, that Audacity will return you an array of 32 ints, if you request sample24 as an output format for the mixer. Am I right, @Paul-Licameli?)

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+
+         char *mixed = (char *)(const char*)mixer->GetBuffer();
+         for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
+            for (size_t i = 0; i < numChannels; i++) {
```
This cycle should be adopted to the different sample formats

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+         }
+
+         if (!WavpackPackSamples(wpc, wavpackBuffer.get(), samplesThisRun))
+            updateResult = ProgressResult::Cancelled;
```
This can be `return ProgressResult::Failed` now, as memory managment is handled by C++ now

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+   }
+
+   if (!WavpackFlushSamples(wpc)) {
+      updateResult = ProgressResult::Cancelled;
```
`return ProgressResult::Failed`?

And I think that `BasicUI::ShowErrorDialog` should be used to notify the user about the error in places where codec has failed. 

(Please, see ExportPCM for example)

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+            n = wxT("DATE");
+         }
+         WavpackAppendTagItem(wpc,
+                              (char *) (const char *) n.mb_str(wxConvUTF8),
```
Both conversions are redundant

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+         }
+         WavpackAppendTagItem(wpc,
+                              (char *) (const char *) n.mb_str(wxConvUTF8),
+                              (char *) (const char *) v.mb_str(wxConvUTF8),
```
Ditto

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+      }
+   }
+
+   WavpackCloseFile(wpc);
```
No longer needed!

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+// src: https://github.com/dbry/WavPack/blob/master/cli/wavpack.c
+int ExportWavPack::WriteBlock(void *id, void *data, int32_t length)
+{
+   WriteId *outId = (WriteId *) id;
```
`static_cast`

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+{
+   WriteId *outId = (WriteId *) id;
+
+   uint32_t bcount;
```
Unused!

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+   uint32_t bcount;
+
+   if (outId->error)
+      return FALSE;
```
`return false;`

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,430 @@
+   if (outId->error)
+      return FALSE;
+
+   if (outId && outId->file && outId->file.get() && data && length) {
```
This function can be rewritten with less nesting involved!

```
int ExportWavPack::WriteBlock(void *id, void *data, int32_t length)
{
    if (id == nullptr || data == nullptr || length == 0)
        return true; // This is considered to be success in wavpack.c reference code

    WriteId *outId = static_cast<WriteId*>(id);

    if (!outId->file)
        // This does not match the wavpack.c but in our case if file is nullptr - 
        // the stream error has occured
        return false; 

    if (!outId->file->Write(data, length).IsOk()) {
        outId->file.reset();
        return false;
    }

    outId->bytesWritten += length;

    if (outId->firstBlockSize == 0)
        outId->firstBlockSize = length;

    return true;
}
```

Also, this allows to remove `error` from the WriteId structure!

crsib reviewed (changes requested):
When I export from Audacity - the resulting file seems to be missing some metadata: VLC does not show the file length and does not allow seeking. Everything works fine, if I generate wv file using the official CLI.

Looks like you miss `WavpackUpdateNumSamples` call in the end.

This concludes the review of the ExportWavPack.cpp, review for the importer will follow.

crsib reviewed (changes requested):
When I export from Audacity - the resulting file seems to be missing some metadata: VLC does not show the file length and does not allow seeking. Everything works fine, if I generate wv file using the official CLI.

Looks like you miss `WavpackUpdateNumSamples` call in the end.

This concludes the review of the ExportWavPack.cpp, review for the importer will follow.

Subhra264 commented:
> When I export from Audacity - the resulting file seems to be missing some metadata: VLC does not show the file length and does not allow seeking. Everything works fine, if I generate wv file using the official CLI.
> 
> Looks like you miss `WavpackUpdateNumSamples` call in the end.

I have replaced `FileIO` with `wxFile` as seeking to the first block of the file was required. I have tested the wavpack files generated from audacity with VLC media player. Now, it seems to work correctly as it is now allowing seeking and the audio length is also shown. Also, I have addressed the changes you asked me to make. The latest commit is now ready for another review.

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -371,21 +413,39 @@ ProgressResult ExportWavPack::Export(AudacityProject *project,
+
+   // wxFile::Create opens the file with only write access
+   // So, need to open the file again with both read and write access
+   if (!outWvFile.file->Open(fName.GetFullPath(), wxFile::read_write)) {
```
For simplicity we can cache that block in memory, but this will do too. Very good!

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -425,6 +486,4 @@ void ExportWavPack::OptionsCreate(ShuttleGui &S, int format)
-};
-
-#endif
+};
```
We generally prefer to have an empty new line at the end of the file

wthrajat replied:
@crsib I am curious to know why :D

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,489 @@
+   ShuttleGui S(this, eIsSavingToPrefs);
+   PopulateOrExchange(S);
+
+   CreateCorrectionFileSetting.Write( mCreateCorrectionFile->GetValue() );
```
Hmm, I'm not sure why this is needed. 

And I think we need to flush settings here

crsib reviewed (commented):
Export looks fine now, I will review import now!

Subhra264 commented:
> Export looks fine now, I will review import now!

Ok then, I will commit and push the remaining ExportWavPack changes along with ImportWavPack changes.

crsib commented on src/import/ImportWavPack.cpp:
```diff
@@ -0,0 +1,292 @@
+
+            // Get the actual length of the value for this item key
+            valueLen = WavpackGetTagItem(mWavPackContext, item, NULL, 0);
+            itemValue = (char *)malloc(valueLen + 1);
```
Ditto

crsib reviewed (changes requested):
Decoding loop need some love :-)

And some minor are changes will be welcome!

Subhra264 commented on src/export/ExportWavPack.cpp:
```diff
@@ -0,0 +1,502 @@
+            const float *mixed = reinterpret_cast<const float*>(mixer->GetBuffer());
+            for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
+               for (size_t i = 0; i < numChannels; i++) {
+                  int64_t intValue = (*mixed++) * (1u << 31);
```
In the `ExportWavPack.cpp` file, I think there were some issues with float 32 bit export. Earlier, I just passed 32 bit int values for float sample export. So, I think while importing it was getting imported as 32 bit int (before adding special cases for float sample import as you suggested). Though it was ok while importing in audacity ( wtih the previous code before applying your suggestions for import), I tried to play the 32 bit exported wavpack file using the VLC media player and it sounded like corrupted. Now, when I converted the `mixed` data into `float` and then into `int32_t` while exporting as 32 bit float, both VLC media and audacity (After applying your suggestions) opened and played the 32 bit float files correctly.

crsib replied:
I think you should set `config->float_norm_exp: = 127;` to enable FP mode in Wavpack. In any case - you should consult `WavpackGetMode` to check, if you need to convert the values.

`1u << 31` - `std::numeric_limits<int32_t>::max()` please.

crsib replied:
(Value `127` is an exponent  bias for IEEE754 32-bit float values. We have 8 bits for exponents, so the bias is `2^7 - 1`)

crsib reviewed (changes requested):
Great work! Just a few more issue, but I think that they are easy to resolve.

crsib commented on src/export/ExportWavPack.cpp:
```diff
@@ -388,7 +389,7 @@ ProgressResult ExportWavPack::Export(AudacityProject *project,
             for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
                for (size_t i = 0; i < numChannels; i++) {
-                  int64_t intValue = (*mixed++) * (1u << 31);
+                  int64_t intValue = (*mixed++) * (std::numeric_limits<int32_t>::max());
```
Minor, but I expect a compiler warning here. Let's add `static_cast<int64_t>` here. Sorry I've missed that before...

crsib commented on src/import/ImportWavPack.cpp:
```diff
@@ -0,0 +1,298 @@
+         samplesRead = WavpackUnpackSamples(mWavPackContext, wavpackBuffer.get(), SAMPLES_TO_READ);
+
+         if (mFormat == int16Sample) {
+            ArrayOf<int16_t> tempBuffer{ bufferSize };
```
Not that important here, but is is best to avoid memory allocations in such loops. Allocations can be quite costly.

crsib reviewed (changes requested):
A one very minor step left!
